M8/advanced-nft/nft.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.7;
+
+import {ERC721} from "@openzeppelin/contracts/token/ERC721/ERC721.sol";
+import {BitMaps} from "@openzeppelin/contracts/utils/structs/BitMaps.sol";
+import {MerkleProof} from "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+
+error AlreadyClaimed();
+error InvalidProof();
+error InvalidStage();
+error NotOwner();
+error InvalidLength();
+error SaleEnded();
+error NotContributor();
+error InvalidMintValue();
+error AlreadyMinted();
+error stillEarlyToMint(uint256 blockNumber);
+error WrongHash();
+error NoMoreContributors();
+
+contract AirDropToken is ERC721 {
+    // bytes32 = [byte, byte, ..., byte] <- 32 bytes
+    bytes32 public immutable merkleRoot;
+    uint256 public mintedNfts;
+    uint256 public constant totalSupply = 1000;
+    address private immutable _owner;
+    BitMaps.BitMap private _whiteList;
+    uint256 public contributorCount;
+    mapping(address => bool) contributors;
+
+    struct Commit {
+        bytes32 commit;
+        uint64 block;
+        bool revealed;
+    }
+    mapping(address => Commit) public commits;
+
+    enum Stages {
+        PendingStart,
+        WhiteList,
+        Public,
+        SaleEnded
+    }
+
+    Stages public stage = Stages.PendingStart;
+
+    event Mint(uint256 tokenId);
+
+    modifier atStage(Stages _stage) {
+        if (stage != _stage) {
+            revert InvalidStage();
+        }
+        _;
+    }
+
+    modifier onlyOwner() {
+        if (msg.sender != _owner) {
+            revert NotOwner();
+        }
+        _;
+    }
+
+    constructor(bytes32 _merkleRoot) ERC721("AirDropNFT", "ADN") {
+        merkleRoot = _merkleRoot;
+        _owner = msg.sender;
+        contributorCount = 1;
+    }
+
+    function addContributor(address contributor) external onlyOwner {
+        if (stage == Stages.SaleEnded) {
+            revert NoMoreContributors();
+        }
+        contributors[contributor] = true;
+        contributorCount++;
+    }
+
+    function NextStage() external onlyOwner {
+        if (stage == Stages.PendingStart) {
+            stage = Stages.WhiteList;
+        } else if (stage == Stages.WhiteList) {
+            stage = Stages.Public;
+        }
+    }
+
+    function claimWhiteList(
+        bytes32[] calldata proof,
+        uint256 index
+    ) external atStage(Stages.WhiteList) {
+        if (stage == Stages.WhiteList) {
+            // check if already claimed
+            if (BitMaps.get(_whiteList, index)) {
+                revert AlreadyClaimed();
+            }
+
+            // verify proof
+            _verifyProof(proof, index, msg.sender);
+
+            // set airdrop as claimed
+            BitMaps.setTo(_whiteList, index, true);
+        }
```

Comment:
* The `Stages.Whitelist` check is not required as `atStage` modifier already checks it
```suggestion
            // check if already claimed
            if (BitMaps.get(_whiteList, index)) {
                revert AlreadyClaimed();
            }

            // verify proof
            _verifyProof(proof, index, msg.sender);

            // set airdrop as claimed
            BitMaps.setTo(_whiteList, index, true);
```